code-generator: add NewFilteredSharedInformerFactory function

munnerz opened the pull request:
**What this PR does / why we need it**:

Adds a `namespace` option to the SharedInformerFactory constructor. This is useful when building controllers that may need to scope themselves to a namespace due to RBAC constraints.

Workarounds for this involve losing type safety if a user wants to use it for core APIs as well as a SharedInformerFactory type interface, as we have to deal with plain SharedIndexInformers (example here: https://github.com/jetstack-experimental/cert-manager/blob/master/pkg/util/kube/factory.go)

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Fixes https://github.com/kubernetes/code-generator/issues/9

**Special notes for your reviewer**:

This will require updating all uses of SharedInformerFactory throughout the codebase. I'm going to follow up with later commits in this PR with these changes, but wanted to get this here to get some feedback on the way it's implemented.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

/cc @sttts @nikhita @deads2k

munnerz committed:
code-generator: add NewFilteredSharedInformerFactory function

Refactor to not change New*Informer constructors

Separate namespace and ListOptions filter

munnerz commented:
/sig api-machinery

cjwagner commented:
/retest

munnerz commented:
This may be better to implement as a `NewNamespacedSharedInformerFactory` function instead of modifying the existing signature to avoid having to touch so many files.

chancez commented:
@munnerz If it's a function, that's not so bad, but I wouldn't want to see a new method added to the interface if thats what you mean.

munnerz commented:
@chancez nope, I mean adding something like:

```
func NewNamespacedSharedInformerFactory(client {{.clientSetInterface|raw}}, namespace string, defaultResync {{.timeDuration|raw}}) SharedInformerFactory {
```

and leaving the existing `NewSharedInformerFactory` function as-is (not changing any interfaces, just add a new constructor method to the package)

chancez commented:
Yeah, I was thinking you may have meant, that. I guess though I'm not sure how that would help, as the namespacing is done by the typed interface that calls the SharedInformerFactories's `InformerFor` method with a non-namespaced `internalinterfaces.NewInformerFunc`. It's the typed Informer that's the issue, not the SharedInformer constructors.

munnerz commented:
I've managed to resolve that by placing the non exported constructor method onto the typed interface. The namespace is plumbed through each group and version interface, along with the factory through each group and versions `New` method.

I've put an example of informers generated with this PR here for you to look at: https://github.com/munnerz/cert-manager/blob/namespaced-informers/third_party/informers/core/v1/configmap.go#L64

You can see here how the namespace is plumbed through to the typed informers: https://github.com/munnerz/cert-manager/blob/namespaced-informers/third_party/informers/factory.go#L121

chancez commented:
@munnerz Yah, I realized it would work, but what about the `Lister` from the informer? We can now pass the namespace to the FooNamespaceLister that may be different from the informer's namespace.

munnerz commented:
Hmm, that's a good point. I don't see how we can work around this then with the current design of the SharedIndexInformer interface/Lister interface/version informer interface.

If a user were to attempt to get or list objects in another namespace, it would just return not found. Perhaps this, combined with it being an additional constructor function (and so not-so invasive), means we could include this?

sttts commented:
/cc @ncdc

sttts commented:
This somehow contradicts the goal of shared informers as this scales with the number of namespaces. We could have a filter method on the existing shared informers which returns an informer filtered by namespace.

wojtek-t commented:
I agree with @sttts - I don't see any good usecase for that. If you have anything specific on your mind I would like to see it.

munnerz commented:
So my use-case for being able to specify a namespace comes from the need to run my application in an environment that constrains the controller to only run in a single namespace. This is a request I've had numerous times, and it usually comes from people running a multi-tenant kubernetes cluster.

Another issue I have, that is similarly related, is that I want to be able to modify the `ListOptions` used on the List & Watch requests. This is especially important with the addition of `Initializers` into Kubernetes, as I need a way to set `IncludeUninitialized: true`. With the current design, I need to create my own implementation of a factory so that I can add this extra field onto the ListOptions. Building on @sttts suggestion, perhaps some way to filter/hook the ListOptions would work - but I'm not sure how we could do this in a clean way.

I think this is probably something that more and more users are going to ask for as people start building out initializers. For now I will probably just fork the informer-gen and create my own informers for core API types, built against `k8s.io/api`/`k8s.io/client-go`. If this is in fact the best way it should be done, we should probably document this on a "build an initializer" page or something similar.

ncdc commented:
Let's take a look at what functions the shared informer factory performs:
1. It creates shared informers on the fly the first time they're referenced in your code
1. It keeps track of which ones you've referenced prior to starting the factory, so it can start them all
1. It allows you to wait for all of its informers' caches to sync
1. It allows you to request a generic informer based on a GroupVersionResource

All of the generated shared informer constructors take in a `namespace` parameter (assuming the resource is namespace-scoped). If all you want is a set of shared informers for a single namespace, you can get that today, albeit without using the factory. You can construct each informer by hand, passing in the desired `namespace`. You can start them as needed. And you can construct the associated lister for each shared informer. BUT, this is stuff you'd have to do by hand, that the factory does for you automatically (assuming you want all namespaces).

Having said all that, I don't think that what @munnerz is asking for is unreasonable, as I've seen multiple people asking for the ability to have a shared informer factory specific to a single namespace. I'd prefer to see it as a new constructor instead of a modification to the existing one (`NewNamespacedSharedInformerFactory`, as described above).

chancez commented:
My use case is the same. My controller is supposed to run with minimal permissions in it's own namespace (no access to non-namespaced resources) and this would greatly reduce the amount of boilerplate i have to write. 

Additionally, I've already ha.d issues where I forgot to add a new informer to my function that waits for caches to sync for all my informers. The autogenerated factories have methods that automatically handle this, and would help prevent mistakes like that.

lavalamp commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:
```diff
@@ -105,9 +106,10 @@ type sharedInformerFactory struct {
 
 // NewSharedInformerFactory constructs a new instance of sharedInformerFactory
-func NewSharedInformerFactory(client {{.clientSetInterface|raw}}, defaultResync {{.timeDuration|raw}}) SharedInformerFactory {
+func NewSharedInformerFactory(client {{.clientSetInterface|raw}}, namespace string, defaultResync {{.timeDuration|raw}}) SharedInformerFactory {
```
Without commenting on whether shared informers should offer this feature or not, I do think these options should be converted to a configuration struct (or use the builder pattern, but that seems unnecessarily complex for this). That way, additions won't break the function signature.

munnerz commented:
I've updated this PR to reflect @sttts suggestion of using a filter function instead of specifying a namespace. I've also switched this to add a new method (`NewFilteredSharedInformerFactory`) instead of modifying the existing one.

As part of that change, a pointer to the `metav1.ListOptions` is also passed to the filter function, which allows users to set fields such as `IncludeUninitialized` (as described in my use case above).

This does not solve the issue of the lister not adhering to the same filter (as @chancez pointed out), but it seems like a good stop gap. We could look at making the generated listers also accept a FilterFunc, but we might then need to consider moving the FilterFunc type into somewhere more generic (e.g. `k8s.io/apimachinery`) in order to avoid having the lister generator depend upon the informer generator, which to me seems slightly backwards. Alternatively we could create the `FilterFunc` interface in the generated listers, and have the generated informers then import that FilterFunc.

Would appreciate hearing feedback on this method!

ncdc commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:
```diff
@@ -107,13 +111,29 @@ type sharedInformerFactory struct {
 // NewSharedInformerFactory constructs a new instance of sharedInformerFactory
 func NewSharedInformerFactory(client {{.clientSetInterface|raw}}, defaultResync {{.timeDuration|raw}}) SharedInformerFactory {
   return &sharedInformerFactory{
```
I'd recommend having this just call `NewFilteredSharedInformerFactory` with `defaultFilterFunc` as the `filter` arg.

ncdc commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/informer.go:
```diff
@@ -123,20 +124,23 @@ type $.type|public$Informer interface {
 // Always prefer using an informer factory to get a shared informer instead of getting an independent
 // one. This reduces memory footprint and number of connections to the server.
-func New$.type|public$Informer(client $.clientSetInterface|raw$$if .namespaced$, namespace string$end$, resyncPeriod $.timeDuration|raw$, indexers $.cacheIndexers|raw$) $.cacheSharedIndexInformer|raw$ {
+func New$.type|public$Informer(client $.clientSetInterface|raw$, filter $.interfacesFilterFunc|raw$, resyncPeriod $.timeDuration|raw$, indexers $.cacheIndexers|raw$) $.cacheSharedIndexInformer|raw$ {
```
This change to the function signature will break anyone calling `New*Informer` when they upgrade their copy of client-go. Would it be possible to find another way to do this that isn't a breaking change? Maybe add another constructor?

ncdc commented:
> As part of that change, a pointer to the metav1.ListOptions is also passed to the filter function, which allows users to set fields such as IncludeUninitialized (as described in my use case above).

This works, but seems a bit counterintuitive. Having said that, I don't have any other ideas atm.

munnerz commented:
@ncdc I've addressed your comments. I've now moved the `defaultFilterFunc` to be `internalinterfaces.DefaultFilterFunc`, and added `internalinterfaces.NamespaceFilter(string) FilterFunc` to make it easier to construct namespace scoped filter functions.

This second function isn't strictly required, but I thought it might make working with this a bit easier 😄 

An example of how this would be used:

```go
exampleInformerFactory := informers.NewFilteredSharedInformerFactory(exampleClient, time.Second*30, filterFunc)

...

func filterFunc(opts *v1.ListOptions) string {
	opts.IncludeUninitialized = true
	// only watch default namespace
	return "default"
}
```

munnerz commented:
Also, I do agree it feels odd to accept one parameter that should be modified (i.e. is a pointer), and return another (namespace).

The alternatives:
- make both parameters be returned, and let the filter function implementer copy things across, but I feel like that's more cumbersome for no gain.
- make the namespace a pointer argument, but that would feel even more awkward to use

So I think we're between a rock and a hard place

munnerz commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/groupinterface.go:
```diff
@@ -98,18 +99,19 @@ type Interface interface {
 // New returns a new Interface.
-func New(f $.interfacesSharedInformerFactory|raw$) Interface {
-	return &group{f}
+func New(f $.interfacesSharedInformerFactory|raw$, filter $.interfacesFilterFunc|raw$) Interface {
```
@ncdc Also worth noting that this changes the `New` constructor for the group and version interface as well, in case anybody is depending upon these directly (although I doubt they are, given it already accepts a SharedInformerFactory as an argument).

ncdc replied:
Yeah, I opted not to comment on this one since it really only exists as an intermediary 😄 .

sttts commented:
@munnerz can you generate an example informer? Would be much easier to comment on the changes than inside of the generator.

munnerz commented:
@sttts I have updated the PR with another commit on top that regenerates the sample-controller informer. I'm trying to get the top-level `update-codegen` working too, so I can update this PR properly.

There *shouldn't* be any breaking changes as a result of the PR so long as nobody has depended upon the `New` function in the group level package, but from it's signature I think that it's unlikely to be used outside of the SharedInformerFactory (as @ncdc mentioned above)

sttts commented on staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/samplecontroller/v1alpha1/foo.go:
```diff
@@ -39,18 +39,29 @@ type FooInformer interface {
 // NewFooInformer constructs a new informer for Foo type.
 // Always prefer using an informer factory to get a shared informer instead of getting an independent
 // one. This reduces memory footprint and number of connections to the server.
 func NewFooInformer(client versioned.Interface, namespace string, resyncPeriod time.Duration, indexers cache.Indexers) cache.SharedIndexInformer {
```
so we had a namespace argument already? What does this PR add then (other than the posibility to tweak the ListOptions) ?

munnerz replied:
Yep we've always had a `namespace` argument on the `NewFooInformer` (or 'leaf' constructors as I like to think of them).

The issue is a few lines below - the `Informer` function (which forms part of the SharedInformerFactory) hardcodes `NamespaceAll`: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/samplecontroller/v1alpha1/foo.go#L64

This means that in order to construct namespaced informers, you cannot use the SharedInformerFactory, which is what the original issue was about 😄 . In order to get around this in the past, I've created a SharedInformerFactory implementation that is non-type aware, and instead just returns SharedIndexInformer instances (I then use a GVK structure as a 'key' for the 'informer caching'. This gets super ugly, and really difficult to keep track of/inconsistent)

sttts commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factoryinterface.go:
```diff
@@ -82,4 +84,16 @@ type SharedInformerFactory interface {
 	InformerFor(obj {{.runtimeObject|raw}}, newFunc NewInformerFunc) {{.cacheSharedIndexInformer|raw}}
 }
+
+type FilterFunc func(*{{.v1ListOptions|raw}}) (namespace string)
```
this is a strange mixture of ListOptions tweaking and filtering a namespace. Also mutation in a filter function is kind of strange.

munnerz replied:
Yep - I'm unsure on the best signature for the FilterFunc. I think allowing users to modify ListOptions *somehow* is reasonable (given things like `IncludeUninitialized`).

What do you think of making the argument passed *to* the FilterFunc be a non-pointer, and expecting the user to return a copy of those ListOptions back (with their modifications, and along with the namespace to operate within as already happens)

sttts replied:
How does the unterlaying store handle uninitialized objects?

munnerz replied:
I've not checked myself, but I'd have thought it'd treat them like any other resource. Upon changes, the store would be updated with the new resource?

From my understanding, uninitialised objects have very little difference to normal objects, except they aren't returned by the apiserver by default (aka without `includeUninitialized=true` on the request) until the `metadata.initializers` field gets set to nil.

sttts commented on staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/samplecontroller/v1alpha1/foo.go:
```diff
@@ -60,12 +71,12 @@ func NewFooInformer(client versioned.Interface, namespace string, resyncPeriod t
 
-func defaultFooInformer(client versioned.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
-	return NewFooInformer(client, v1.NamespaceAll, resyncPeriod, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+func (f *fooInformer) defaultInformer(client versioned.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
```
What if we change `NewInformerFunc` with an additional namespace arg? That would be the canonical change without the need to tweak the list watcher last minute with a filter, wouldn't it?

munnerz replied:
I'm not sure I understand - you mean change the `defaultInformer` (or `defaultFooInformer` previously) to have a `namespace` argument?

The issue there, is that `InformerFor` in the `SharedInformerFactory` then needs to somehow be aware whether the Informer that it's constructing is for a namespaced type or not. We *could* just include `namespace` on *all* of them, and silently drop the namespace arg on non-namespaced resources.

This sort of goes against what you mentioned previously - that we could use a filter instead of just a plain string. I'd *hope* I'm not alone in wanting to be able to modify `ListOptions` of the informers I create, but perhaps I am!

sttts replied:
> We could just include namespace on all of them, and silently drop the namespace arg on non-namespaced resources.

That's what I had in mind. But you have a point. What is really ugly is the filter func doing two things. You could separate the namespace and the filter by:
- keep the namespace string separated in `NewFilteredFooInformer`, store it in `group` as well and pass it to `NewFilteredFooInformer` from `defaultInformer`
- rename filter to tweakListOptions.

It's not really elegant either, but less magic.

munnerz replied:
Yeah that might be a better idea - I imagine a lot of people will want to set a namespace, but be okay with the default ListOptions. Do you think that `tweakListOptions` function should accept a pointer, or a copy of the type to then return? My only concern then was that a user should really perform a DeepCopy before modifying the copy passed into the filter, as ListOptions could at some point include some nested/pointer types. Perhaps this isn't a concern though. By passing a pointer, it's clear straight away that this type *will* be mutated.

I guess the 3rd option is to create some kind of 'Filter builder', where we can pass in a list of arbitrary functions that manipulate some kind of 'Filter' structure, that is understood by the `New*Informer` functions. This will probably be a fair amount more work, and may not be *clearer* to the user, but it'd probably be more elegant.

munnerz replied:
(the 3rd option there obviously does not help fix the Lister inconsistency. I think we really are stuck on that one without a big change)

sttts replied:
> Do you think that tweakListOptions function should accept a pointer

`tweakListOptions` is explicit enough to expect side effects. So a pointer is fine.

munnerz commented:
Another thing:

When using an Informer produced from a factory with a filter set, if a user attempts to use the lister to get/list resources, they may find a 404 of 0 list items found is returned as the cache has not been filled with items from the specified namespace/matching the filters on ListOptions.

Short of a major refactoring of the lister interface, I don't think we've really got another way around this. Perhaps a comment on the `NewFilteredSharedInformerFactory` function to explain this would suffice. wdyt @ncdc @sttts

sttts commented:
> don't think we've really got another way around this

that's what I feared.

munnerz committed:
run hack/update-codegen.sh

munnerz commented:
@sttts I've updated to include a `TweakListOptionsFunc` instead of `FilterFunc`, and once again made `namespace` a field on `NewFilteredSharedInformerFactory`.

PTAL and let me know what you think. It definitely does feel a bit more sane now!

munnerz committed:
run hack/update-bazel.sh

munnerz commented:
/retest

munnerz commented:
Unit tests appear to be failing, I'm not certain whether it's related although I really think it isn't given if it *were* down to this PR, I'd imagine a tonne more tests would be failing 😄 

/retest

sttts commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:
```diff
@@ -106,8 +110,15 @@ type sharedInformerFactory struct {
   return &sharedInformerFactory{
-		client: client,
+	client:           client,
+    namespace:        namespace,
```
indention?

munnerz replied:
Fixed - mixture between spaces and tabs confused my IDE!

munnerz commented:
Rebased onto master.

@sttts @ncdc @lavalamp is this something that can be accepted? I'm using this in some of my own projects now with success 😄

ncdc commented:
You're back to where it doesn't modify any of the preexisting function signatures that developers would tend to invoke, right?

munnerz commented:
@ncdc Yep. There are two changes to exported functions:

* The constructor for the group interface: https://github.com/kubernetes/kubernetes/blob/9cb0cdeca9d72d0a452fb74ea282ede8fc21154b/pkg/client/informers/informers_generated/internalversion/core/interface.go#L39

* The constructor for the version interface: https://github.com/kubernetes/kubernetes/blob/9cb0cdeca9d72d0a452fb74ea282ede8fc21154b/pkg/client/informers/informers_generated/internalversion/core/internalversion/interface.go#L68

Aside from that, it adds the following:

* `NewFilteredSharedInformerFactory(client internalclientset.Interface, defaultResync time.Duration, namespace string, tweakListOptions internalinterfaces.TweakListOptionsFunc) SharedInformerFactory` - https://github.com/kubernetes/kubernetes/blob/9cb0cdeca9d72d0a452fb74ea282ede8fc21154b/pkg/client/informers/informers_generated/internalversion/factory.go#L65

* `NewFilteredPodInformer(client internalclientset.Interface, namespace string, resyncPeriod time.Duration, indexers cache.Indexers, tweakListOptions internalinterfaces.TweakListOptionsFunc)` - https://github.com/kubernetes/kubernetes/blob/9cb0cdeca9d72d0a452fb74ea282ede8fc21154b/pkg/client/informers/informers_generated/internalversion/core/internalversion/pod.go#L56

munnerz commented:
/retest

ncdc commented:
I'm fine with proceeding.

On Tue, Nov 7, 2017 at 5:45 PM, James Munnelly <notifications@github.com>
wrote:

> /retest
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/54660#issuecomment-342649033>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYsc5qdiCJk0GvEyOhs5gW3b9-9EUks5s0N1wgaJpZM4QIUuC>
> .
>

sttts commented:
So the only downside is the Lister inconsistency. Is this true? Can we document this at least?

munnerz commented:
@sttts yep exactly. Although this isn't the first instance of this inconsistency - previously if you were to have called `NewPodInformer` and specified a namespace other than NamespaceAll, you would also see this same behaviour (404 for anything outside of that namespace).

Agreed that documentation is best. Where do you think this is best to go? I'd guess in a comment in the generated file? Perhaps a comment on `NewFilteredSharedInformerFactory` would be clearest?

munnerz commented:
@sttts I've updated the PR to include a comment on the generated `NewFilteredSharedInformerFactory` function clarifying the behaviour.

sttts commented:
As discussed on Slack with @munnerz the inconsistency is actually in the opposite direction, i.e. the listers have the same filters applied as the informer. IMO this is not inconsistent at all, but the expected behaviour. So this is fine with me.

sttts commented on staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:
```diff
@@ -109,8 +113,17 @@ type sharedInformerFactory struct {
+}
+
+// NewFilteredSharedInformerFactory constructs a new instance of sharedInformerFactory.
+// Listers obtained via this SharedInformerFactory will be subject to the same filters
```
:+1:

sttts commented:
lgtm

sttts commented:
/lgtm

sttts commented:
/approve no-issue

munnerz commented:
/retest

munnerz commented:
/retest

munnerz commented:
/retest

munnerz commented:
@lavalamp @sttts tests are now passing again after problems with Prow and a few flakes.

If you could take a final look over and approve I'd greatly appreciate!

munnerz commented:
@lavalamp @sttts more conflicts came up which I have once again resolved. Looks like this is going to happen a lot as this PR touches `pkg/client`!

Hopefully test flakes won't hold this up too much again 😄 but this will need another /lgtm since rebasing!

sttts commented:
/lgtm

sttts commented:
@smarterclayton are you fine with this?

deads2k commented:
/approve
